File: dashlane-project/Dashlane/src/main/java/com/dashlane/teamspaces/manager/TeamspaceManager.java
package com.dashlane.teamspaces.manager;

import android.content.Context;

import com.dashlane.R;
import com.dashlane.dagger.singleton.SingletonProvider;
import com.dashlane.logger.ExceptionLog;
import com.dashlane.logger.Log;
import com.dashlane.preference.UserPreferencesManager;
import com.dashlane.settings.SettingsManager;
import com.dashlane.teamspaces.CombinedTeamspace;
import com.dashlane.teamspaces.PersonalTeamspace;
import com.dashlane.teamspaces.db.TeamspaceForceCategorizationManager;
import com.dashlane.teamspaces.db.TeamspaceUsageLogSpaceChanged;
import com.dashlane.teamspaces.model.Teamspace;
import com.dashlane.util.Constants;
import com.dashlane.util.StringUtils;
import com.dashlane.util.ThreadHelper;
import com.dashlane.util.userfeatures.UserFeaturesChecker;

import org.jetbrains.annotations.NotNull;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.Set;

import androidx.annotation.ColorRes;
import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.annotation.VisibleForTesting;
import androidx.collection.ArrayMap;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;



public class TeamspaceManager implements TeamspaceAccessor {

    private static final String LOG_TAG = "TEAMSPACE";

    public static final Teamspace PERSONAL_TEAMSPACE = PersonalTeamspace.INSTANCE;
    public static final Teamspace COMBINED_TEAMSPACE = CombinedTeamspace.INSTANCE;
    private static final Teamspace DEFAULT_TEAMSPACE = COMBINED_TEAMSPACE;

    private final SpaceAnonIdDataProvider mDataProvider;

    private final List<Teamspace> mJoinedTeamspaces = new ArrayList<>();
    private final List<Teamspace> mRevokedAndDeclinedSpaces = new ArrayList<>();
    private final Set<Listener> mListeners = new HashSet<>();

    private Teamspace mCurrent = DEFAULT_TEAMSPACE;
    private TeamspaceRestrictionNotificator mTeamspaceNotifier;
    private SettingsManager mSettingsManager;

    public TeamspaceManager(SpaceAnonIdDataProvider spaceAnonIdDataProvider,
                            SettingsManager settingsManager,
                            List<Teamspace> teamspaces,
                            TeamspaceUsageLogSpaceChanged teamspaceUsageLogSpaceChanged) {
        subscribeListener(new RevokedDetector());
        mSettingsManager = settingsManager;
        mDataProvider = spaceAnonIdDataProvider;
        initEmbeddedTeamspace();
        mTeamspaceNotifier = new TeamspaceRestrictionNotificator();
        subscribeListener(teamspaceUsageLogSpaceChanged);
    }

    public final void init(List<Teamspace> data, TeamspaceForceCategorizationManager forceCategorizationManager) {

        List<Teamspace> previousJoinedTeamspaces = getAll();
        List<Teamspace> previousRevokedAndDeclineTeamspaces = new ArrayList<>(mRevokedAndDeclinedSpaces);

        mJoinedTeamspaces.clear();
        mRevokedAndDeclinedSpaces.clear();

        
        mJoinedTeamspaces.add(0, COMBINED_TEAMSPACE);
        mJoinedTeamspaces.add(1, PERSONAL_TEAMSPACE);

        for (int i = 0; data != null && i < data.size(); i++) {
            Teamspace t = data.get(i);
            String status = t.getStatus();
            if (status == null) {
                continue;
            }
            switch (status) {
                case Teamspace.Status.DECLINED:
                case Teamspace.Status.REVOKED:
                    mRevokedAndDeclinedSpaces.add(t);
                    break;
                case Teamspace.Status.ACCEPTED:
                    mJoinedTeamspaces.add(t);
                    break;
            }
        }

        SpaceAnonIdDataProvider dataProvider = getDataProvider();
        setupAnalyticsIds(dataProvider);

        reloadCurrentSpace();

        

        notifyAllUpdateListeners();

        notifyChangeStatus(previousJoinedTeamspaces, previousRevokedAndDeclineTeamspaces);

        forceCategorizationManager.executeAsync();
    }

    

    public SpaceAnonIdDataProvider getDataProvider() {
        return mDataProvider;
    }

    

    @NonNull
    @Override
    public List<Teamspace> getAll() {
        List<Teamspace> allJoinedTeamspaces = new ArrayList<>(mJoinedTeamspaces);
        List<Teamspace> nonNullTeamspaces = new ArrayList<>(allJoinedTeamspaces.size());
        for (Teamspace teamspace : allJoinedTeamspaces) {
            if (teamspace != null) {
                nonNullTeamspaces.add(teamspace);
            }
        }
        return nonNullTeamspaces;
    }

    

    @NonNull
    @Override
    public List<Teamspace> getRevokedAndDeclinedSpaces() {
        return mRevokedAndDeclinedSpaces;
    }

    @Override
    public Teamspace get(@NonNull String id) {
        Teamspace teamspace = getTeamspace(getAll(), id);
        if (teamspace != null) return teamspace;
        return getTeamspace(mRevokedAndDeclinedSpaces, id);
    }

    @Override
    public boolean isCurrent(@NonNull String id) {
        return !canChangeTeamspace() 
               || (mCurrent != null && Objects.equals(mCurrent.getTeamId(), id));
    }

    

    @Override
    public Teamspace getCurrent() {
        return mCurrent;
    }

    @NotNull
    @Override
    public Teamspace getCombinedTeamspace() {
        return COMBINED_TEAMSPACE;
    }

    

    @Override
    public void setCurrent(Teamspace teamspace) {
        if (!Teamspace.Status.ACCEPTED.equals(teamspace.getStatus())) {
            return;
        }
        mCurrent = teamspace;
        setAsDefault(teamspace);

        notifyAllChangeListeners();
    }

    

    public boolean isSpaceSelected() {
        return !COMBINED_TEAMSPACE.equals(getCurrent());
    }

    

    @Override
    public boolean canChangeTeamspace() {
        return getAll().size() > 2;
    }

    

    public final void subscribeListener(Listener listener) {
        mListeners.add(listener);
    }

    

    public void unSubscribeListeners(@Nullable Listener listener) {
        mListeners.remove(listener);
    }

    @VisibleForTesting
    void setupAnalyticsIds(SpaceAnonIdDataProvider dataProvider, SettingsManager settingsManager) {
        List<Teamspace> teamspaces = new ArrayList<>();
        teamspaces.addAll(getAll());
        teamspaces.addAll(mRevokedAndDeclinedSpaces);
        ArrayMap<Teamspace, String> anonSpaceMap = dataProvider.getAnonSpaceIds(settingsManager, teamspaces);
        for (int i = 0; i < teamspaces.size(); i++) {
            Teamspace space = teamspaces.get(i);
            String anonId = anonSpaceMap.get(space);
            space.setAnonTeamId(anonId);
        }
    }

    private void setupAnalyticsIds(SpaceAnonIdDataProvider dataProvider) {
        setupAnalyticsIds(dataProvider, mSettingsManager);
    }

    private void notifyAllUpdateListeners() {
        if (!ThreadHelper.isMainThread()) {
            SingletonProvider.getThreadHelper().runOnMainThread(() -> notifyAllUpdateListeners());
            return;
        }
        Set<Listener> listeners = new HashSet<>(mListeners);
        for (Listener reference : listeners) {
            reference.onTeamspacesUpdate();
        }
    }

    private void notifyAllChangeListeners() {
        if (!ThreadHelper.isMainThread()) {
            SingletonProvider.getThreadHelper().runOnMainThread(() -> notifyAllChangeListeners());
            return;
        }
        Set<Listener> listeners = new HashSet<>(mListeners);
        for (Listener reference : listeners) {
            reference.onChange(mCurrent);
        }
    }

    private void setAsDefault(@NonNull Teamspace teamspace) {
        try {
            JSONObject jsonDescriptor = new JSONObject();
            jsonDescriptor.put(Constants.DEFAULT_TEAMSPACE_TYPE, teamspace.getType());
            jsonDescriptor.put(Constants.DEFAULT_TEAMSPACE_ID, teamspace.getTeamId());

            SingletonProvider.getUserPreferencesManager().putString(Constants.DEFAULT_TEAMSPACE,
                                                                    jsonDescriptor.toString());
        } catch (JSONException e) {
            Log.v(LOG_TAG, "failed to store default space, json descriptor failed to create", e);
        }
    }

    

    private void reloadCurrentSpace() {
        mCurrent = getDefaultSpaceType();
    }

    @NonNull
    private Teamspace getDefaultSpaceType() {
        UserPreferencesManager preferencesManager = SingletonProvider.getUserPreferencesManager();
        try {
            if (!preferencesManager.exist(Constants.DEFAULT_TEAMSPACE)) {
                return DEFAULT_TEAMSPACE;
            }
            String descriptor = preferencesManager.getString(Constants.DEFAULT_TEAMSPACE);
            JSONObject jsonDescriptor = new JSONObject(descriptor);
            int type = jsonDescriptor.getInt(Constants.DEFAULT_TEAMSPACE_TYPE);
            String id = jsonDescriptor.optString(Constants.DEFAULT_TEAMSPACE_ID);
            switch (type) {
                case Teamspace.Type.PERSONAL:
                    return PERSONAL_TEAMSPACE;
                case Teamspace.Type.COMBINED:
                    return COMBINED_TEAMSPACE;
                case Teamspace.Type.COMPANY:
                    for (Teamspace space : getAll()) {
                        if (id != null && id.equals(space.getTeamId())) {
                            return space;
                        }
                    }
                    return DEFAULT_TEAMSPACE;
                default:
                    return DEFAULT_TEAMSPACE;
            }
        } catch (JSONException e) {
            Log.v(LOG_TAG, "failed to store default space, json descriptor failed to create", e);
        }
        return DEFAULT_TEAMSPACE;


    }

    private void initEmbeddedTeamspace() {
        init(PERSONAL_TEAMSPACE, R.string.teamspace_personal, R.color.teamspace_personal);
        init(COMBINED_TEAMSPACE, R.string.teamspace_combined, R.color.teamspace_combined);
    }

    private void init(Teamspace teamspace, @StringRes int labelResId, @ColorRes int colorResId) {
        Context context = SingletonProvider.getContext();
        if (context == null) {
            return;
        }
        String label = context.getString(labelResId);
        int color = ContextCompat.getColor(context, colorResId);
        teamspace.setTeamName(label);
        teamspace.setCompanyName(label);
        teamspace.setDisplayLetter(label.substring(0, 1));
        teamspace.setColor(String.format(Locale.US, "#%06X", (0xFFFFFF & color)));
    }

    

    @Override
    public boolean isFeatureEnabled(@Teamspace.Feature String featureCheck) {
        if (!canChangeTeamspace()) {
            
            return true;

        }
        if (!StringUtils.isNotSemanticallyNull(featureCheck)) {
            
            return true;
        }
        for (Teamspace space : getAll()) {
            
            if (space.getType() == Teamspace.Type.PERSONAL ||
                space.getType() == Teamspace.Type.COMBINED) {
                
                continue;
            }
            if (space.featureDisabledForSpace(featureCheck)) {
                return false;
            }
        }
        return true;
    }

    

    @Nullable
    @Override
    public String getFeatureValue(@NonNull @Teamspace.Feature String featureCheck) {
        if (!canChangeTeamspace()) {
            
            return null;

        }
        if (!StringUtils.isNotSemanticallyNull(featureCheck)) {
            
            return null;
        }
        for (Teamspace space : getAll()) {
            
            if (space.getType() == Teamspace.Type.PERSONAL ||
                space.getType() == Teamspace.Type.COMBINED) {
                
                continue;
            }
            return space.getFeatureValue(featureCheck);
        }
        return null;
    }

    private void notifyChangeStatus(List<Teamspace> previousJoinedTeamspaces,
                                    List<Teamspace> previousRevokedAndDeclineTeamspaces) {
        if (previousJoinedTeamspaces.size() == 0) {
            
            return;
        }
        List<Teamspace> allTeamspaces = getAll();
        for (int i = 2; i < allTeamspaces.size(); i++) { 
            Teamspace teamspace = allTeamspaces.get(i);
            String previousStatus;
            if (isPresent(previousJoinedTeamspaces, teamspace)) {
                continue; 
            } else if (isPresent(previousRevokedAndDeclineTeamspaces, teamspace)) {
                previousStatus = Teamspace.Status.REVOKED;
            } else {
                previousStatus = null;
            }
            sendStatusUpdate(teamspace, previousStatus);
        }
        for (int i = 0; i < mRevokedAndDeclinedSpaces.size(); i++) {
            Teamspace teamspace = mRevokedAndDeclinedSpaces.get(i);
            String previousStatus;
            if (isPresent(previousJoinedTeamspaces, teamspace)) {
                previousStatus = Teamspace.Status.ACCEPTED;
            } else {
                Teamspace previousSpace = getTeamspace(previousRevokedAndDeclineTeamspaces, teamspace.getTeamId());
                if (previousSpace == null) {
                    previousStatus = null;
                } else if (previousSpace.shouldDeleteForceCategorizedContent() ==
                           teamspace.shouldDeleteForceCategorizedContent()) {
                    continue; 
                } else {
                    previousStatus = Teamspace.Status.REVOKED;
                }
            }
            sendStatusUpdate(teamspace, previousStatus);
        }
    }

    private void sendStatusUpdate(final Teamspace teamspace, final String previousStatus) {
        if (!ThreadHelper.isMainThread()) {
            SingletonProvider.getThreadHelper().runOnMainThread(() -> sendStatusUpdate(teamspace, previousStatus));
            return;
        }
        Set<Listener> listeners = new HashSet<>(mListeners);
        for (Listener reference : listeners) {
            reference.onStatusChanged(teamspace, previousStatus, teamspace.getStatus());
        }
    }

    @Nullable
    private Teamspace getTeamspace(List<Teamspace> teamspaces, String id) {
        for (int i = 0, size = teamspaces.size(); i < size; i++) {
            Teamspace teamspace = teamspaces.get(i);
            if (Objects.equals(teamspace.getTeamId(), id)) {
                return teamspace;
            }
        }
        return null;
    }

    private boolean isPresent(List<Teamspace> list, Teamspace teamspace) {
        return getTeamspace(list, teamspace.getTeamId()) != null;
    }

    

    public interface Listener {
        

        @MainThread
        void onStatusChanged(Teamspace teamspace, String previousStatus, String newStatus);

        

        @MainThread
        void onChange(Teamspace teamspace);

        

        @MainThread
        void onTeamspacesUpdate();
    }

    @Override
    public void startFeatureOrNotify(FragmentActivity activity,
                                     @Teamspace.Feature String feature,
                                     FeatureCall featureCall) {
        if (Teamspace.Feature.SECURE_NOTES_DISABLED.equals(feature)
            && SingletonProvider.getUserFeatureChecker().has(UserFeaturesChecker.FeatureFlip.DISABLE_SECURE_NOTES)) {
            mTeamspaceNotifier.notifyFeatureRestricted(activity, feature);
            return;
        }

        if (!isFeatureEnabled(feature)) {
            mTeamspaceNotifier.notifyFeatureRestricted(activity, feature);
            return;
        }
        featureCall.startFeature();
    }
}

File: dashlane-project/Dashlane/src/main/java/com/dashlane/teamspaces/manager/TeamspaceManager.kt
package com.dashlane.teamspaces.manager

import androidx.annotation.ColorRes
import androidx.annotation.MainThread
import androidx.annotation.StringRes
import androidx.annotation.VisibleForTesting
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import com.dashlane.R
import com.dashlane.dagger.singleton.SingletonProvider
import com.dashlane.settings.SettingsManager
import com.dashlane.teamspaces.CombinedTeamspace
import com.dashlane.teamspaces.PersonalTeamspace
import com.dashlane.teamspaces.db.TeamspaceForceCategorizationManager
import com.dashlane.teamspaces.db.TeamspaceUsageLogSpaceChanged
import com.dashlane.teamspaces.manager.TeamspaceAccessor.FeatureCall
import com.dashlane.teamspaces.model.Teamspace
import com.dashlane.util.Constants
import com.dashlane.util.ThreadHelper
import com.dashlane.util.isNotSemanticallyNull
import com.dashlane.util.userfeatures.UserFeaturesChecker
import org.json.JSONException
import org.json.JSONObject
import java.util.Locale

class TeamspaceManager(
    val dataProvider: SpaceAnonIdDataProvider,
    private val settingsManager: SettingsManager,
    teamspaceUsageLogSpaceChanged: TeamspaceUsageLogSpaceChanged?
) : TeamspaceAccessor {
    private val defaultTeamspace = CombinedTeamspace

    private val joinedTeamspaces: MutableList<Teamspace> = ArrayList()
    private val _revokedAndDeclinedSpaces: MutableList<Teamspace> = ArrayList()
    private val listeners: MutableSet<Listener> = HashSet()

    override var current: Teamspace? = defaultTeamspace
        set(teamspace) {
            if (Teamspace.Status.ACCEPTED != teamspace?.status) {
                return
            }
            field = teamspace
            setAsDefault(teamspace)
            notifyAllChangeListeners()
        }

    private val teamspaceNotifier = TeamspaceRestrictionNotificator()

    init {
        subscribeListener(RevokedDetector())
        initEmbeddedTeamspace()
        subscribeListener(teamspaceUsageLogSpaceChanged)
    }

    fun init(data: List<Teamspace>?, forceCategorizationManager: TeamspaceForceCategorizationManager) {
        val previousJoinedTeamspaces = all
        val previousRevokedAndDeclineTeamspaces: List<Teamspace> = ArrayList(_revokedAndDeclinedSpaces)
        joinedTeamspaces.clear()
        _revokedAndDeclinedSpaces.clear()

        
        joinedTeamspaces.add(0, CombinedTeamspace)
        joinedTeamspaces.add(1, PersonalTeamspace)
        var i = 0
        while (data != null && i < data.size) {
            val t = data[i]
            val status = t.status
            if (status == null) {
                i++
                continue
            }
            when (status) {
                Teamspace.Status.DECLINED, Teamspace.Status.REVOKED -> _revokedAndDeclinedSpaces.add(t)
                Teamspace.Status.ACCEPTED -> joinedTeamspaces.add(t)
            }
            i++
        }
        val dataProvider = dataProvider
        setupAnalyticsIds(dataProvider)
        reloadCurrentSpace()

        
        notifyAllUpdateListeners()
        notifyChangeStatus(previousJoinedTeamspaces, previousRevokedAndDeclineTeamspaces)
        forceCategorizationManager.executeAsync()
    }

    override val all: List<Teamspace>
        get() = joinedTeamspaces.filterNotNull().toList()

    override val revokedAndDeclinedSpaces: List<Teamspace>
        get() = _revokedAndDeclinedSpaces.filterNotNull().toList()

    override operator fun get(id: String): Teamspace? {
        val teamspace = getTeamspace(all, id)
        return teamspace ?: getTeamspace(_revokedAndDeclinedSpaces, id)
    }

    override fun isCurrent(id: String): Boolean {
        return !canChangeTeamspace() || current != null && current!!.teamId == id
    }

    val isSpaceSelected: Boolean
        get() = CombinedTeamspace != current

    override fun canChangeTeamspace(): Boolean {
        return all.size > 2
    }

    fun subscribeListener(listener: Listener?) {
        listener?.let {
            listeners.add(it)
        }
    }

    fun unSubscribeListeners(listener: Listener?) {
        listeners.remove(listener)
    }

    @VisibleForTesting
    fun setupAnalyticsIds(dataProvider: SpaceAnonIdDataProvider, settingsManager: SettingsManager?) {
        val teamspaces: MutableList<Teamspace> = ArrayList()
        teamspaces.addAll(all)
        teamspaces.addAll(_revokedAndDeclinedSpaces)
        val anonSpaceMap = dataProvider.getAnonSpaceIds(settingsManager, teamspaces)
        for (i in teamspaces.indices) {
            val space = teamspaces[i]
            val anonId = anonSpaceMap[space]
            space.anonTeamId = anonId
        }
    }

    private fun setupAnalyticsIds(dataProvider: SpaceAnonIdDataProvider) {
        setupAnalyticsIds(dataProvider, settingsManager)
    }

    private fun notifyAllUpdateListeners() {
        if (!ThreadHelper.isMainThread()) {
            SingletonProvider.getThreadHelper().runOnMainThread { notifyAllUpdateListeners() }
            return
        }
        val listeners: Set<Listener> = HashSet(listeners)
        for (reference in listeners) {
            reference.onTeamspacesUpdate()
        }
    }

    private fun notifyAllChangeListeners() {
        if (!ThreadHelper.isMainThread()) {
            SingletonProvider.getThreadHelper().runOnMainThread { notifyAllChangeListeners() }
            return
        }
        val listeners: Set<Listener> = HashSet(listeners)
        for (reference in listeners) {
            reference.onChange(current)
        }
    }

    private fun setAsDefault(teamspace: Teamspace) {
        try {
            val jsonDescriptor = JSONObject()
            jsonDescriptor.put(Constants.DEFAULT_TEAMSPACE_TYPE, teamspace.type)
            jsonDescriptor.put(Constants.DEFAULT_TEAMSPACE_ID, teamspace.teamId)
            SingletonProvider.getUserPreferencesManager().putString(
                Constants.DEFAULT_TEAMSPACE,
                jsonDescriptor.toString()
            )
        } catch (e: JSONException) {
            warn("failed to store default space, json descriptor failed to create", LOG_TAG, e)
        }
    }

    private fun reloadCurrentSpace() {
        current = defaultSpaceType
    }

    private val defaultSpaceType: Teamspace
        get() {
            val preferencesManager = SingletonProvider.getUserPreferencesManager()
            try {
                if (!preferencesManager.exist(Constants.DEFAULT_TEAMSPACE)) {
                    return defaultTeamspace
                }
                val descriptor = preferencesManager.getString(Constants.DEFAULT_TEAMSPACE)
                val jsonDescriptor = JSONObject(descriptor)
                val type = jsonDescriptor.getInt(Constants.DEFAULT_TEAMSPACE_TYPE)
                val id = jsonDescriptor.optString(Constants.DEFAULT_TEAMSPACE_ID)
                return when (type) {
                    Teamspace.Type.PERSONAL -> PersonalTeamspace
                    Teamspace.Type.COMBINED -> CombinedTeamspace
                    Teamspace.Type.COMPANY -> {
                        for (space in all) {
                            if (id != null && id == space.teamId) {
                                return space
                            }
                        }
                        defaultTeamspace
                    }
                    else -> defaultTeamspace
                }
            } catch (e: JSONException) {
                warn("failed to store default space, json descriptor failed to create", LOG_TAG, e)
            }
            return defaultTeamspace
        }

    private fun initEmbeddedTeamspace() {
        init(PersonalTeamspace, R.string.teamspace_personal, R.color.teamspace_personal)
        init(CombinedTeamspace, R.string.teamspace_combined, R.color.teamspace_combined)
    }

    private fun init(teamspace: Teamspace, @StringRes labelResId: Int, @ColorRes colorResId: Int) {
        val context = SingletonProvider.getContext() ?: return
        val label = context.getString(labelResId)
        val color = ContextCompat.getColor(context, colorResId)
        teamspace.teamName = label
        teamspace.companyName = label
        teamspace.displayLetter = label.substring(0, 1)
        teamspace.color = String.format(Locale.US, "#%06X", 0xFFFFFF and color)
    }

    override fun isFeatureEnabled(@Teamspace.Feature featureCheck: String): Boolean {
        if (!canChangeTeamspace()) {
            
            return true
        }
        if (!featureCheck.isNotSemanticallyNull()) {
            
            return true
        }
        for (space in all) {
            
            if (space.type == Teamspace.Type.PERSONAL ||
                space.type == Teamspace.Type.COMBINED
            ) {
                
                continue
            }
            if (space.featureDisabledForSpace(featureCheck)) {
                return false
            }
        }
        return true
    }

    override fun getFeatureValue(@Teamspace.Feature featureCheck: String): String? {
        if (!canChangeTeamspace()) {
            
            return null
        }
        if (!featureCheck.isNotSemanticallyNull()) {
            
            return null
        }
        for (space in all) {
            
            if (space.type == Teamspace.Type.PERSONAL ||
                space.type == Teamspace.Type.COMBINED
            ) {
                
                continue
            }
            return space.getFeatureValue(featureCheck)
        }
        return null
    }

    private fun notifyChangeStatus(
        previousJoinedTeamspaces: List<Teamspace>,
        previousRevokedAndDeclineTeamspaces: List<Teamspace>
    ) {
        if (previousJoinedTeamspaces.isEmpty()) {
            
            return
        }
        val allTeamspaces = all

        for (teamspace in allTeamspaces.drop(2)) { 
            val previousStatus: String? = if (isPresent(previousJoinedTeamspaces, teamspace)) {
                continue 
            } else if (isPresent(previousRevokedAndDeclineTeamspaces, teamspace)) {
                Teamspace.Status.REVOKED
            } else {
                null
            }
            sendStatusUpdate(teamspace, previousStatus)
        }
        for (teamspace in _revokedAndDeclinedSpaces) {
            val previousStatus: String? = if (isPresent(previousJoinedTeamspaces, teamspace)) {
                Teamspace.Status.ACCEPTED
            } else {
                val previousSpace = getTeamspace(previousRevokedAndDeclineTeamspaces, teamspace.teamId)
                if (previousSpace == null) {
                    null
                } else if (previousSpace.shouldDeleteForceCategorizedContent() ==
                    teamspace.shouldDeleteForceCategorizedContent()
                ) {
                    continue 
                } else {
                    Teamspace.Status.REVOKED
                }
            }
            sendStatusUpdate(teamspace, previousStatus)
        }
    }

    private fun sendStatusUpdate(teamspace: Teamspace, previousStatus: String?) {
        if (!ThreadHelper.isMainThread()) {
            SingletonProvider.getThreadHelper().runOnMainThread { sendStatusUpdate(teamspace, previousStatus) }
            return
        }
        val listeners: Set<Listener> = HashSet(listeners)
        for (reference in listeners) {
            reference.onStatusChanged(teamspace, previousStatus, teamspace.status)
        }
    }

    private fun getTeamspace(teamspaces: List<Teamspace>, id: String?) = teamspaces.firstOrNull { it.teamId == id }

    private fun isPresent(list: List<Teamspace>, teamspace: Teamspace): Boolean {
        return getTeamspace(list, teamspace.teamId) != null
    }

    interface Listener {
        @MainThread
        fun onStatusChanged(teamspace: Teamspace?, previousStatus: String?, newStatus: String?)

        @MainThread
        fun onChange(teamspace: Teamspace?)

        @MainThread
        fun onTeamspacesUpdate()
    }

    override fun startFeatureOrNotify(
        activity: FragmentActivity?,
        @Teamspace.Feature feature: String?,
        featureCall: FeatureCall?
    ) {
        if (Teamspace.Feature.SECURE_NOTES_DISABLED == feature && SingletonProvider.getUserFeatureChecker()
                .has(UserFeaturesChecker.FeatureFlip.DISABLE_SECURE_NOTES)
        ) {
            teamspaceNotifier.notifyFeatureRestricted(activity, feature)
            return
        }
        if (!isFeatureEnabled(feature!!)) {
            teamspaceNotifier.notifyFeatureRestricted(activity, feature)
            return
        }
        featureCall!!.startFeature()
    }

    companion object {
        private const val LOG_TAG = "TEAMSPACE"
    }
}